homeassistant/components/timer.py

```diff
@@ -0,0 +1,315 @@
+    vol.Optional(ATTR_DURATION,
+                 default=timedelta(DEFAULT_DURATION)): cv.time_period,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        cv.slug: vol.Any({
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_ICON): cv.icon,
+            vol.Optional(CONF_DURATION, timedelta(DEFAULT_DURATION)):
+                cv.time_period,
+        }, None)
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+@bind_hass
+def sync_start(hass, entity_id, duration):
+    """Start a timer."""
+    hass.add_job(async_start, hass, entity_id, {ATTR_ENTITY_ID: entity_id,
+                                                ATTR_DURATION: duration})
+
+
+@callback
+@bind_hass
+def async_start(hass, entity_id, duration):
+    """Start a timer."""
+    hass.async_add_job(hass.services.async_call(
+        DOMAIN, SERVICE_START, {ATTR_ENTITY_ID: entity_id,
+                                ATTR_DURATION: duration}))
+
+
+@bind_hass
+def pause(hass, entity_id):
+    """Pause a timer."""
+    hass.add_job(async_pause, hass, entity_id)
+
+
+@callback
+@bind_hass
+def async_pause(hass, entity_id):
+    """Pause a timer."""
+    hass.async_add_job(hass.services.async_call(
+        DOMAIN, SERVICE_PAUSE, {ATTR_ENTITY_ID: entity_id}))
+
+
+@bind_hass
+def cancel(hass, entity_id):
+    """Cancel a timer."""
+    hass.add_job(async_cancel, hass, entity_id)
+
+
+@callback
+@bind_hass
+def async_cancel(hass, entity_id):
+    """Cancel a timer."""
+    hass.async_add_job(hass.services.async_call(
+        DOMAIN, SERVICE_CANCEL, {ATTR_ENTITY_ID: entity_id}))
+
+
+@bind_hass
+def finish(hass, entity_id):
+    """Finish a timer."""
+    hass.add_job(async_cancel, hass, entity_id)
+
+
+@callback
+@bind_hass
+def async_finish(hass, entity_id):
+    """Finish a timer."""
+    hass.async_add_job(hass.services.async_call(
+        DOMAIN, SERVICE_FINISH, {ATTR_ENTITY_ID: entity_id}))
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up a timer."""
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    entities = []
+
+    for object_id, cfg in config[DOMAIN].items():
+        if not cfg:
+            cfg = {}
+
+        name = cfg.get(CONF_NAME)
+        icon = cfg.get(CONF_ICON)
+        duration = cfg.get(CONF_DURATION)
+
+        entities.append(Timer(hass, object_id, name, icon, duration))
+
+    if not entities:
+        return False
+
+    @asyncio.coroutine
+    def async_handler_service(service):
+        """Handle a call to the timer services."""
+        target_timers = component.async_extract_from_service(service)
+
+        attr = None
+        if service.service == SERVICE_PAUSE:
+            attr = 'async_pause'
+        elif service.service == SERVICE_CANCEL:
+            attr = 'async_cancel'
+        elif service.service == SERVICE_FINISH:
+            attr = 'async_finish'
+
+        tasks = [getattr(timer, attr)() for timer in target_timers if attr]
+        if service.service == SERVICE_START:
```

Comment:
For a future PR, we could consider adding the ability to define timers on the fly from a service and have them remove themselves when done.